File: app/src/main/java/com/mkf_test/showtexts/ListActivity.java
package com.mkf_test.showtexts;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.ListView;
import android.widget.TextView;

import com.mkf_test.showtexts.db.table.WebUrl;
import com.mkf_test.showtexts.utils.Dbutils;

import java.util.List;

import androidx.appcompat.app.AlertDialog;
import butterknife.BindView;

public class ListActivity extends BaseActivity {
    @BindView(R.id.list)
    ListView listView;
    List<WebUrl> webUrllist;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initData();
    }

    @Override
    public int getLayoutResId() {
        return R.layout.activity_list;
    }

    private void initData() {
        setTitle("书签");

        try {
            webUrllist = Dbutils.getInstance().dbGetList();
            if (webUrllist!=null)
            listView.setAdapter(new MyAdapter(this,webUrllist));
        } catch (Exception e) {
            e.printStackTrace();
        }

        listView.setOnItemClickListener(new AdapterView.OnItemClickListener(){

            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                openUrl(i);
            }
        });
        listView.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(final AdapterView<?> parent, final View view, final int position, long id) {
                new AlertDialog.Builder(ListActivity.this)
                        .setItems(R.array.longclickmsg,
                                new DialogInterface.OnClickListener() {
                                    public void onClick(DialogInterface dialog,
                                                        int which) {
                                        switch (which){
                                            case 0://打开
                                                openUrl(position);
                                                break;
                                            case 1://复制网址
                                                ClipboardManager cm = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
                                                cm. setPrimaryClip(ClipData.newPlainText(null, webUrllist.get(position).getUrl()));
                                                break;
                                            case 2://删除
                                                delectUrl(position);
                                                ((MyAdapter)listView.getAdapter()).notifyDataSetChanged();
                                                break;
                                        }

                                    }
                                })
                        .show();
                return true;
            }
        });

    }
    private  void openUrl(int position){
        if (webUrllist!=null) {
            setResult(RESULT_OK, getIntent().putExtra("url", webUrllist.get(position).getUrl()));
            finish();
        }
    }
    private  void delectUrl(int position){
        if (webUrllist!=null) {
            try {
                Dbutils.getInstance().dbDelete(webUrllist.get(position));
                webUrllist.remove(position);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    class MyAdapter extends BaseAdapter {
        Context context; List<WebUrl> webUrllist;
        private MyAdapter(Context context, List<WebUrl> webUrllist){
            this.context=context;
            this.webUrllist=webUrllist;
        }
        @Override

        public int getCount() {
            return webUrllist.size();
        }

        @Override
        public Object getItem(int i) {
            return null;
        }

        @Override
        public long getItemId(int i) {
            return 0;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup group) {
            ViewHolder holder = null;
            if (convertView == null) {
                holder = new ViewHolder();
                convertView = LayoutInflater.from(context).inflate(
                       R.layout.adapter_item_text, group, false);
                holder.name= (TextView) convertView.findViewById(R.id.text1);
                holder.name.setTextColor(getResources().getColor(R.color.black));
                convertView.setTag(holder);
            } else {
                holder = (ViewHolder) convertView.getTag();
            }
            holder.setPosition(position);
            return convertView;
        }

        class ViewHolder {
            TextView name;

            public void setPosition(final int position) {
                name.setText(webUrllist.get(position).getName());
            }
        }
    }
}

File: app/src/main/java/com/mkf_test/showtexts/ListActivity.kt
package com.mkf_test.showtexts

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.BaseAdapter
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import com.mkf_test.showtexts.db.query.WebUrlTableQuery
import com.mkf_test.showtexts.db.table.WebUrlTable
import kotlinx.android.synthetic.main.activity_list.*

class ListActivity : BaseActivity() {
    //    @BindView(R.id.list)
//    internal var listView: ListView? = null
    private var webUrlList: MutableList<WebUrlTable>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initData()
    }

    override fun getLayoutResId(): Int {
        return R.layout.activity_list
    }

    private fun initData() {
        title = "书签"

        try {
            webUrlList = WebUrlTableQuery().dbGetList()
            if (webUrlList != null)
                list?.adapter = MyAdapter(this, webUrlList as MutableList<WebUrlTable>)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        list.onItemClickListener = AdapterView.OnItemClickListener { adapterView, view, i, l -> openUrl(i) }
        list.onItemLongClickListener = AdapterView.OnItemLongClickListener { parent, view, position, id ->
            AlertDialog.Builder(this@ListActivity)
                    .setItems(R.array.longclickmsg
                    ) { dialog, which ->
                        when (which) {
                            0//打开
                            -> openUrl(position)
                            1//复制网址
                            -> {
                                val cm = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                                cm.setPrimaryClip(ClipData.newPlainText(null, webUrlList!![position].url))
                            }
                            2//删除
                            -> {
                                delectUrl(position)
                                (list.adapter as MyAdapter).notifyDataSetChanged()
                            }
                        }
                    }
                    .show()
            true
        }

    }

    private fun openUrl(position: Int) {
        if (webUrlList != null) {
            setResult(RESULT_OK, intent.putExtra("url", webUrlList!![position].url))
            finish()
        }
    }

    private fun delectUrl(position: Int) {
        if (webUrlList != null) {
            try {
                WebUrlTableQuery().deleteObject(webUrlList!![position])
                webUrlList!!.removeAt(position)
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
    }

    internal inner class MyAdapter(var context: Context, var webUrllist: List<WebUrlTable>) : BaseAdapter() {
        override fun getCount(): Int {
            return webUrllist.size
        }

        override fun getItem(i: Int): Any? {
            return null
        }

        override fun getItemId(i: Int): Long {
            return 0
        }

        override fun getView(position: Int, convertView: View?, group: ViewGroup): View {
            val view: View
            val holder: ViewHolder
            if (convertView == null) {
                holder = ViewHolder()
                view = LayoutInflater.from(context).inflate(
                        R.layout.adapter_item_text, group, false)
                holder.name = view?.findViewById<TextView>(R.id.text1)
                holder.name?.setTextColor(resources.getColor(R.color.black))
                view.tag = holder
            } else {
                view = convertView
                holder = view.tag as ViewHolder
            }
            holder.setPosition(position)
            return view
        }

        internal inner class ViewHolder {
            var name: TextView? = null

            fun setPosition(position: Int) {
                name?.text = webUrllist[position].getName()
            }
        }
    }
}
